### PR TITLE
[SPARK-6119][SQL] DataFrame support for missing data handling

### PR DESCRIPTION
**Pull request by rxin**
This pull request adds variants of DataFrame.na.drop and DataFrame.na.fill to the Scala/Java API, and DataFrame.fillna and DataFrame.dropna to the Python API.

### CODE REVIEW

**Comment by rxin** [comment]
This subsumes https://github.com/apache/spark/pull/5248

File: python/pyspark/sql/dataframe.py [hunk 1]
```diff
@@ -697,6 +697,83 @@ def subtract(self, other):
...
+        """
+        if subset is None:
+            subset = self.columns
+        elif isinstance(subset, basestring):
+            subset = [subset]
+        elif not isinstance(subset, (list, tuple)):
+            raise ValueError("subset should be a list or tuple of column names")
+
+        if thresh is None:
+            thresh = len(subset) if how == 'any' else 1
```
Comment (davies) [inline python/pyspark/sql/dataframe.py]: how should be one of `any` or `all`

Comment (rxin) [inline python/pyspark/sql/dataframe.py]: did you mean adding an explicit check?

Comment (davies) [inline python/pyspark/sql/dataframe.py]: yes, or all others will become `all`

Comment (rxin) [inline python/pyspark/sql/dataframe.py]: fixed.

File: python/pyspark/sql/dataframe.py [hunk 2]
```diff
@@ -697,6 +697,83 @@ def subtract(self, other):
...
+        10  80     Alice
+        5   null   Bob
+        50  null   Tom
+        50  null   unknown
+        """
+        if not isinstance(value, (float, int, long, basestring, dict)):
+            raise ValueError("value should be a float, int, long, string, or dict")
+
+        if isinstance(value, (int, long)):
+            value = float(value)
```
Comment (davies) [inline python/pyspark/sql/dataframe.py]: Why we use `float` instead of `int` here?

**Comment by davies** [comment]
LGTM.

File: python/pyspark/sql/dataframe.py (hunk 2 above)
Comment (rxin) [inline python/pyspark/sql/dataframe.py]: the jvm implementation only accepts a double

Comment (davies) [inline python/pyspark/sql/dataframe.py]: I see.

**Comment by rxin** [comment]
Merging in master & branch-1.3. We will need to formalize the semantics of null vs NaN and come back and update the documentation later.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+   * Returns a new [[DataFrame]] that drops rows containing any null values
+   * in the specified columns.
+   */
+  def drop(cols: Array[String]): DataFrame = drop(cols.toSeq)
+
+  /**
+   * (Scala-specific) Returns a new [[DataFrame ]] that drops rows containing any null values
+   * in the specified columns.
+   */
+  def drop(cols: Seq[String]): DataFrame = drop(cols.size, cols)
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala]: Why have this and the `Array[String]` version?  `Array`s get auto converted to `Seq`.

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala]: Java

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala]: My point is why not only have the java friendly `Array` version?

**Pull request closed without merging**